File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoMiddleware/Bedrock.swift
import Foundation
import IAIsoCore

/// AWS Bedrock runtime middleware. Supports Converse API (preferred —
/// normalized usage extraction) and the lower-level InvokeModel API.
public enum Bedrock {

    public protocol Client {
        func converse(_ params: [String: Any]) throws -> ConverseResponse
        func invokeModel(_ params: [String: Any]) throws -> InvokeResponse
    }

    public struct ConverseUsage: Sendable, Equatable {
        public let inputTokens: Int64
        public let outputTokens: Int64
        public let totalTokens: Int64
        public init(
            inputTokens: Int64 = 0,
            outputTokens: Int64 = 0,
            totalTokens: Int64 = 0
        ) {
            self.inputTokens = inputTokens
            self.outputTokens = outputTokens
            self.totalTokens = totalTokens
        }
    }

    public struct ConverseContentBlock: Sendable, Equatable {
        public let hasToolUse: Bool
        public init(hasToolUse: Bool = false) { self.hasToolUse = hasToolUse }
    }

    public struct ConverseResponse: Sendable {
        public let usage: ConverseUsage
        public let content: [ConverseContentBlock]
        public init(usage: ConverseUsage, content: [ConverseContentBlock]) {
            self.usage = usage
            self.content = content
        }
    }

    public struct InvokeResponse: Sendable {
        public let modelId: String
        public let body: Data
        public init(modelId: String = "", body: Data = Data()) {
            self.modelId = modelId
            self.body = body
        }
    }

    public struct Options: Sendable {
        public var raiseOnEscalation: Bool
        public init(raiseOnEscalation: Bool = false) {
            self.raiseOnEscalation = raiseOnEscalation
        }
        public static let defaults = Options()
    }

    public final class BoundedClient {
        private let raw: Client
        private let execution: BoundedExecution
        private let opts: Options

        public init(raw: Client, execution: BoundedExecution, options: Options = .defaults) {
            self.raw = raw
            self.execution = execution
            self.opts = options
        }

        public func converse(_ params: [String: Any]) throws -> ConverseResponse {
            try checkState()
            let resp: ConverseResponse
            do {
                resp = try raw.converse(params)
            } catch {
                throw MiddlewareError.provider(
                    error.localizedDescription, underlying: error)
            }
            var tokens = resp.usage.totalTokens
            if tokens == 0 {
                tokens = resp.usage.inputTokens + resp.usage.outputTokens
            }
            let toolCalls: Int64 = resp.content.reduce(0) { acc, b in
                acc + (b.hasToolUse ? 1 : 0)
            }
            let modelId = (params["modelId"] as? String) ?? "unknown"
            execution.recordStep(StepInput(
                tokens: tokens,
                toolCalls: toolCalls,
                tag: "bedrock.converse:\(modelId)"))
            return resp
        }

        public func invokeModel(_ params: [String: Any]) throws -> InvokeResponse {
            try checkState()
            let resp: InvokeResponse
            do {
                resp = try raw.invokeModel(params)
            } catch {
                throw MiddlewareError.provider(
                    error.localizedDescription, underlying: error)
            }
            let modelId = !resp.modelId.isEmpty
                ? resp.modelId : ((params["modelId"] as? String) ?? "unknown")
            execution.recordStep(StepInput(tag: "bedrock.invokeModel:\(modelId)"))
            return resp
        }

        private func checkState() throws {
            let pre = execution.check()
            if pre == .locked { throw MiddlewareError.locked }
            if pre == .escalated, opts.raiseOnEscalation {
                throw MiddlewareError.escalationRaised
            }
        }
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/middleware/src/bedrock.rs
//! AWS Bedrock runtime wrapper. Supports both Converse (preferred —
//! normalized usage extraction across model families) and InvokeModel.

use crate::MiddlewareError;
use iaiso_core::{BoundedExecution, StepInput, StepOutcome};
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ConverseUsage {
    #[serde(default)]
    pub input_tokens: u64,
    #[serde(default)]
    pub output_tokens: u64,
    #[serde(default)]
    pub total_tokens: u64,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ConverseToolUse {
    #[serde(default)]
    pub tool_use_id: String,
    #[serde(default)]
    pub name: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ConverseContentBlock {
    #[serde(default)]
    pub tool_use: Option<ConverseToolUse>,
    #[serde(default)]
    pub text: String,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct ConverseMessage {
    #[serde(default)]
    pub content: Vec<ConverseContentBlock>,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct ConverseOutput {
    #[serde(default)]
    pub message: ConverseMessage,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ConverseResponse {
    #[serde(default)]
    pub output: ConverseOutput,
    #[serde(default)]
    pub usage: ConverseUsage,
}

#[derive(Debug, Clone)]
pub struct InvokeResponse {
    pub model_id: String,
    pub body: Vec<u8>,
}

pub trait Client: Send + Sync {
    fn converse(&self, params: &Value) -> Result<ConverseResponse, String>;
    fn invoke_model(&self, params: &Value) -> Result<InvokeResponse, String>;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Options {
    pub raise_on_escalation: bool,
}

pub struct BoundedClient<'a, C: Client> {
    raw: C,
    execution: &'a BoundedExecution,
    opts: Options,
}

impl<'a, C: Client> BoundedClient<'a, C> {
    pub fn new(raw: C, execution: &'a BoundedExecution, opts: Options) -> Self {
        Self { raw, execution, opts }
    }

    pub fn converse(&self, params: &Value) -> Result<ConverseResponse, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self.raw.converse(params).map_err(MiddlewareError::Provider)?;
        let mut tokens = resp.usage.total_tokens;
        if tokens == 0 {
            tokens = resp.usage.input_tokens + resp.usage.output_tokens;
        }
        let tool_calls = resp
            .output
            .message
            .content
            .iter()
            .filter(|b| b.tool_use.is_some())
            .count() as u64;
        let model_id = params
            .get("modelId")
            .and_then(Value::as_str)
            .unwrap_or("unknown")
            .to_string();
        let _ = self.execution.record_step(StepInput {
            tokens,
            tool_calls,
            tag: Some(format!("bedrock.converse:{}", model_id)),
            ..Default::default()
        });
        Ok(resp)
    }

    pub fn invoke_model(&self, params: &Value) -> Result<InvokeResponse, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self
            .raw
            .invoke_model(params)
            .map_err(MiddlewareError::Provider)?;
        let model_id = if !resp.model_id.is_empty() {
            resp.model_id.clone()
        } else {
            params
                .get("modelId")
                .and_then(Value::as_str)
                .unwrap_or("unknown")
                .to_string()
        };
        let tokens = extract_tokens_from_invoke_body(&resp.body);
        let _ = self.execution.record_step(StepInput {
            tokens,
            tag: Some(format!("bedrock.invokeModel:{}", model_id)),
            ..Default::default()
        });
        Ok(resp)
    }
}

/// Best-effort scan of a model-specific InvokeModel response body for
/// token counts. Returns 0 when the body shape is unrecognized.
pub fn extract_tokens_from_invoke_body(_body: &[u8]) -> u64 {
    // Operators who need precise accounting on InvokeModel should adapt
    // this for their specific model family. The Converse wrapper above
    // handles usage uniformly.
    0
}
